### ISSUE TITLE
uiform question: how could a new form called within a form?

### ISSUE THREAD

**Issue by karliwalti** [part-cw/lambdanative#189]
I'd like to call a form directly within a form if a certain condition is met. 
E.g. like in the obviously broken example below (main is not a uiform element), where the desired behavior would be that whenever next is  pressed in main the form page is loaded but jumps back to main.
I'd like to avoid pressing a button where main is the action

Any ideas how to achieve this?

```
(main 
"" #f ("next" form)
(spacer)
)
(form
"" #f #f 
,(lambda () (if #t `(main) ))
)
```

**Comment by ddunsmuir** [comment]
I'm not sure I follow your request. You want to be able to press "next" (the top right button) on the main page and have this go to another page called form, but then have that redirect back to the main page - or more straightforward, you just want the next button to refresh the current page (main)?

The page is automatically refreshed pretty often - whenever any actions are taken on it, so I'm not sure what the extra refresh is for? What do you want to change on the current page?

**Comment by karliwalti** [comment]
Correct. A redirection would be desirable if the form has nothing to answer (eg branching logic.)
The redirection to main was just an example, it could be also another page.
What makes a redirection necessary is that in branching logic, the previous page does not know the next's logic.
Best

**Comment by ddunsmuir** [comment]
Ah, ok. I have added a redirect element. Please see commit [251949](https://github.com/part-cw/lambdanative/commit/25194914e1767964e1a38f0681b18ab956fd73dc).

The below script will then do what you want:

(main 
"" #f ("next" form)
(spacer)
)
(form
"" #f #f 
(redirect action ,(lambda () 'main))
)

**Comment by karliwalti** [comment]
Dear Dustin

Great idea to add it there, thank you so much.

**Comment by karliwalti** [comment]
I'm observing a strange behavior that I cannot explain:

```
`( 
(main
"main" #f ("next" form)
(spacer)
(button text "main" action ,(lambda () '("main" ( "OK" #f))))
)
(form
"form" #f #f
(redirect action ,(lambda () 'main))
(button text "form" action ,(lambda () '("form" ( "OK" #f))))
)
)
```
After pressing next, the first callback of the main button will be from form, then it is back to the main callback.  Is action somwhere retained in memory? I can't find anything that would indicate this.

**Comment by ddunsmuir** [comment]
Hmm, I have been able to reproduce this. Perhaps it has to do with the redirect action being done from the draw procedure, not the input procedure. I'll look into it.

**Comment by ddunsmuir** [comment]
Yes, it was due to this happening during draw. The clearing of the nodemap needs to happen after the full draw of the page is completed. This is fixed in [e183140](https://github.com/part-cw/lambdanative/commit/e1831400a0c6dafe44c497014eac4271de1fe94a). You should not need to update your script. It should just work properly now.
